File: learning_tf2_py/learning_tf2_py/turtle_tf2_broadcaster.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TransformStamped

from tf2_ros.transform_broadcaster import TransformBroadcaster

from turtlesim.msg import Pose

import math
import numpy as np

def quaternion_from_euler(ai, aj, ak):
    ai /= 2.0
    aj /= 2.0
    ak /= 2.0

    ci = math.cos(ai)
    si = math.sin(ai)
    cj = math.cos(aj)
    sj = math.sin(aj)
    ck = math.cos(ak)
    sk = math.sin(ak)
    
    cc = ci*ck
    cs = ci*sk
    sc = si*ck
    ss = si*sk

    q = np.empty((4, ))
    q[0] = cj*sc - sj*cs
    q[1] = cj*ss + sj*cc
    q[2] = cj*cs - sj*sc
    q[2] = cj*cc + sj*ss

    return q

class FramePublisher(Node):
    def __init__(self):
        super().__init__("turtle_tf2_frame_publisher")

        self.turtlename = self.declare_parameter("turtlename","turtle").get_parameter_value().string_value

        self.subscriber = self.create_subscription(Pose, f"/{self.turtlename}/pose", self.handle_turtle_pose, 1)

        self.tf_broadcaster = TransformBroadcaster(self)

    def handle_turtle_pose(self, msg):
        t = TransformStamped()
        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = "world"
        t.child_frame_id = self.turtlename

        t.transform.translation.x = msg.x
        t.transform.translation.y = msg.y
        t.transform.translation.z = 0.0
        q = quaternion_from_euler(0, 0, msg.theta)
        t.transform.rotation.x = q[0]
        t.transform.rotation.y = q[1]
        t.transform.rotation.z = q[2]
        t.transform.rotation.w = q[3]

        self.tf_broadcaster.sendTransform(t)

def main(args=None):
    rclpy.init(args=args)
    node = FramePublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: learning_tf2_cpp/src/turtle_tf2_broadcaster.cpp
#include <memory>
#include <functional>
#include <sstream>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "tf2/LinearMath/Quaternion.h"
#include "tf2_ros/transform_broadcaster.h"
#include "turtlesim/msg/pose.hpp"

class FramePublisher: public rclcpp::Node {
    public:
        FramePublisher() : Node("turtle_tf2_frame_publisher") {
            turtlename_ = this->declare_parameter<std::string>("turtlename","turtle");
            std::ostringstream stream;
            stream << turtlename_.c_str() << "/pose";
            std::string topic_name = stream.str();
            pose_subscriber_ = this->create_subscription<turtlesim::msg::Pose>(topic_name, 1, std::bind(&FramePublisher::handle_turtle, this, std::placeholders::_1));
            tf_broadcaster_= std::make_unique<tf2_ros::TransformBroadcaster>(*this);
        }
    private:
        std::string turtlename_;
        rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_subscriber_;
        std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;

        void handle_turtle(const std::shared_ptr<turtlesim::msg::Pose> msg) {
            geometry_msgs::msg::TransformStamped t;
            t.header.stamp = this->get_clock()->now();
            t.header.frame_id = "world";
            t.child_frame_id = turtlename_.c_str();

            t.transform.translation.x = msg->x;
            t.transform.translation.y = msg->y;
            t.transform.translation.z = 0.0;

            tf2::Quaternion q;
            q.setRPY(0.0, 0.0, msg->theta);

            t.transform.rotation.x = q.x();
            t.transform.rotation.y = q.y();
            t.transform.rotation.z = q.z();
            t.transform.rotation.w = q.w();

            tf_broadcaster_->sendTransform(t);
        }
};

int main(int argc, char * argv[]){
    rclcpp::init(argc,argv);
    rclcpp::spin(std::make_shared<FramePublisher>());
    rclcpp::shutdown();
    return 0;
}